### PR TITLE
gcstar: init at 1.7.2 + perl package dependencies

### PR DESCRIPTION
**Pull request by dasj19**
###### Motivation for this change

This is my first package made from scratch. So please point out eventual beginner errors. Thanks

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/gcstar/default.nix [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    perlPackages.GDGraph
+    perlPackages.GDText
+    perlPackages.HTMLParser
+    perlPackages.librelative
+    perlPackages.LWPUserAgent
+    perlPackages.Pango
+    perlPackages.XMLSimple
+    perlPackages.XMLParser ];
+
+  buildInputs = [makeWrapper] ++ perlDeps;
```
Comment (symphorien) [inline pkgs/applications/misc/gcstar/default.nix]: `makeWrapper` should be a nativeBuildInput

File: pkgs/applications/misc/gcstar/default.nix [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    description = "GCstar is an application for managing your collections.";
+    longDescription = ''
+    GCstar is an application for managing your collections.
+    It supports many types of collections, including movies, books, games, comics, stamps, coins, and many more.
+    You can even create your own collection type for whatever unique thing it is that you collect!
+    Detailed information on each item can be automatically retrieved from the internet and you can store additional data, such as the location or who you've lent it to. You may also search and filter your collections by many criteria.
+    '';
+    license = licenses.gpl2;
+    platforms = platforms.all;
+    outputsToInstall = [ "out" ];
```
Comment (symphorien) [inline pkgs/applications/misc/gcstar/default.nix]: I think this is unnecessary, as their is only one output.

File: pkgs/applications/misc/gcstar/default.nix [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    sed -i 's|/usr/bin/perl|'"${perl}"'/bin/perl|g' bin/gcstar
+    sed -i 's|/usr/bin/perl|'"${perl}"'/bin/perl|g' ./install
+    sed -i 's|/bin/sh|'"${bash}"'/bin/bash|g' share/gcstar/helpers/xdg-open
+
+    perl install --text --prefix=$out
+  '';
+
+  postFixup = "wrapProgram $out/bin/gcstar --prefix PERL5LIB : $PERL5LIB";
+
+  outputs = [ "out" ];
```
Comment (symphorien) [inline pkgs/applications/misc/gcstar/default.nix]: This is the default.

File: pkgs/applications/misc/gcstar/default.nix [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  meta = with stdenv.lib; {
+    homepage = https://launchpad.net/gcstar;
+    description = "GCstar is an application for managing your collections.";
```
Comment (symphorien) [inline pkgs/applications/misc/gcstar/default.nix]: As per https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md, `meta.description` should:
    * Be capitalized.
    * Not start with the package name.
    * Not have a period at the end.

Comment (dasj19) [reply pkgs/applications/misc/gcstar/default.nix]: @symphorien thanks for review... I fixed your points and added handling for a desktop file and its icon

**Comment by symphorien** [comment]
When I run it, after selecting a collection type, the UI stays frozen:
```
***   undef is not of type Gtk2::Widget at /nix/store/vgx575rq53c02vvryw8k4pzxvhdv2689-gcstar-1.7.1/bin/../lib/gcstar/GCGraphicComponents/GCBaseWidgets.pm line 1013.
```

Maybe you need wrapGappsHook ?

**Comment by dasj19** [comment]
@symphorien I only made it fail selecting a music collection type so far... which type did you select?

**Comment by symphorien** [comment]
Wine

**Comment by dasj19** [comment]
@symphorien I found that the code from original gcstar project appears to no longer be maintained but someone continued the work other place: https://gitlab.com/Kerenoc/GCstar/

So I applied those updates and fixed all the perl package dependencies... and now it should work with all types of collections

File: pkgs/applications/misc/gcstar/default.nix [hunk 5]
```diff
@@ -0,0 +1,82 @@
+{stdenv, fetchurl, bash, perl, gtk2, perlPackages, wrapGAppsHook}:
+
+stdenv.mkDerivation rec {
+  pname = "gcstar";
+  version = "1.7.2";
+
+  src = fetchurl {
```
Comment (symphorien) [inline pkgs/applications/misc/gcstar/default.nix]: any specific reason to not use `fetchFromGitHub` ?

File: pkgs/applications/misc/gcstar/default.nix [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  buildInputs = perlDeps;
+  propagatedBuildInputs = perlDeps;
+
+  installPhase = ''
+    tar xvfz $src
+    cd GCstar-v${version}/gcstar
+
+    sed -i 's|/usr/bin/perl|'"${perl}"'/bin/perl|g' bin/gcstar
+    sed -i 's|/usr/bin/perl|'"${perl}"'/bin/perl|g' ./install
+    sed -i 's|/bin/sh|'"${bash}"'/bin/bash|g' share/gcstar/helpers/xdg-open
```
Comment (symphorien) [inline pkgs/applications/misc/gcstar/default.nix]: use patchShebangs instead. For executables in `$out` it's automatic.

File: pkgs/top-level/perl-packages.nix [hunk 7]
```diff
@@ -13613,6 +13664,21 @@ let
...
   };
 
+  NetFreeDB = buildPerlPackage {
+    pname = "Net-FreeDB";
+    version = "0.10";
+    src = fetchurl {
+      url = mirror://cpan/authors/id/D/DS/DSHULTZ/Net-FreeDB-0.10.tar.gz;
+      sha256 = "11dfi14qnzsnmr71cygir85zfj15n08b7d5g0i4cj5pb70if2hzp";
+    };
+    doCheck = false;
```
Comment (symphorien) [inline pkgs/top-level/perl-packages.nix]: When you leave `doCheck = false` can you leave a comment of what does not work ?

**Comment by symphorien** [comment]
Yes now it works.

I think you can simplify the derivation by using patchShebangs (it is automatically ran after the installPhase) and fetchFromGitHub, which unpacks the source for you.

**Review by symphorien — APPROVED** [review]

**Comment by dasj19** [comment]
@symphorien thanks for approving.. I also fixed the doCheck issues for the perl modules

**Comment by dasj19** [comment]
In an attempt to squash the 7 git commits I had and update to master I kinda messed up. Please feel free to cleanup the git history (I don't want to mess it up anymore).
And this package has been sitting waiting for a merge for quite some time. Any reasons for not merging?

**Comment by jonringer** [comment]
`git rebase --onto origin/master HEAD~7` that should take the last 7 commits and put them on top of master. Not sure how merge conflicts do or don't factor into this, so take the command with a grain of salt

**Comment by dasj19** [comment]
> `git rebase --onto origin/master HEAD~7` that should take the last 7 commits and put them on top of master. Not sure how merge conflicts do or don't factor into this, so take the command with a grain of salt

Thanks for advice @jonringer unfortunately that did not work but the following approach seems to work: https://stackoverflow.com/a/53336320/5073290 even when you have merges in-between actual commits

**Comment by dasj19** [comment]
`nix path-info -S /nix/store/1nh35y1bxs9vh9pxgiy3sxn6hvn09dns-gcstar-1.7.2 -h` reports 863.3MB

**Comment by dasj19** [comment]
@jtojnar Can we get this old PR merged?

File: pkgs/applications/misc/gcstar/default.nix [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+stdenv.mkDerivation rec {
+  pname = "gcstar";
+  version = "1.7.2";
+
+  src = fetchFromGitLab {
+      owner = "Kerenoc";
+      repo = "GCstar";
+      rev = "v${version}";
+      sha256 = "1vqfff33sssvlvsva1dflggmwl00j5p64sn1669f9wrbvjkxgpv4";
+    };
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:12]: Indentation.

File: pkgs/applications/misc/gcstar/default.nix [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+  enableParallelBuilding = false;
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:14]: This should be off by default unless you use cmake/meson/qmake.

File: pkgs/applications/misc/gcstar/default.nix [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    perlPackages.librelative
+    perlPackages.LWPUserAgent
+    perlPackages.LWPProtocolHttps
+    perlPackages.MP3Info
+    perlPackages.MP3Tag
+    perlPackages.NetFreeDB
+    perlPackages.OggVorbisHeaderPurePerl
+    perlPackages.Pango
+    perlPackages.XMLSimple
+    perlPackages.XMLParser ];
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:38]: Could you move this out of the `mkDerivation` using a `let` expression? (Assuming this is only here to deduplicate `{,propagated}BuildInputs`.

Comment (jtojnar) [reply pkgs/applications/misc/gcstar/default.nix:38]: Though, maybe just `buildInputs` or just `propagatedBuildInputs` would be enough. Not familiar with Perl packaging enough to be able to say which.

File: pkgs/applications/misc/gcstar/default.nix [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+  nativeBuildInputs = [ wrapGAppsHook ];
+  buildInputs = perlDeps;
+  propagatedBuildInputs = perlDeps;
+
+  installPhase = ''
+    cd gcstar
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:45]: When overriding `installPhase`, you should also provide `runHook preInstall` and `runHook postInstall`.

File: pkgs/applications/misc/gcstar/default.nix [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+    perl install --text --prefix=$out
+
+    # Setting up the desktop item.
+    mkdir -p $out/share/applications
+    mkdir -p $out/share/pixmaps
+    cp share/applications/$pname.desktop $out/share/applications/$pname.desktop
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:52]: Is not that installed by default? https://gitlab.com/Kerenoc/GCstar/-/blob/dde0f1fbcc3c5b6bd68307ed2c3af0d68ea24bdf/gcstar/install#L182

File: pkgs/applications/misc/gcstar/default.nix [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    cp share/$pname/icons/${pname}_256x256.png $out/share/pixmaps/$pname.png
+  '';
+
+  postFixup = ''
+    patchShebangs bin/gcstar
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:57]: This should be done automatically by `patchShebangsAuto`.

File: pkgs/applications/misc/gcstar/default.nix [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    patchShebangs ./install
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:58]: Why is this needed? By this time, the installation script already ran.

File: pkgs/applications/misc/gcstar/default.nix [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+    wrapProgram $out/bin/gcstar --prefix PERL5LIB : $PERL5LIB
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = https://gitlab.com/Kerenoc/GCstar;
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:64]: Unquoted URLs are being phased out: https://github.com/NixOS/rfcs/pull/45

File: pkgs/applications/misc/gcstar/default.nix [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    description = "A general purpose collection manager";
+    longDescription = ''
+    GCstar is an application for managing your collections.
+    It supports many types of collections, including movies, books, games, comics, stamps, coins, and many more.
+    You can even create your own collection type for whatever unique thing it is that you collect!
+    Detailed information on each item can be automatically retrieved from the internet and you can store additional data, such as the location or who you've lent it to. You may also search and filter your collections by many criteria.
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:70]: Missing indentation.

File: pkgs/applications/misc/gcstar/default.nix [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    '';
+    license = licenses.gpl2;
+    platforms = platforms.all;
+  };
+  
+  inherit perl;
+  inherit (perlPackages) Gtk2;
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:77]: Why is this needed?

File: pkgs/top-level/perl-packages.nix [hunk 18]
```diff
@@ -13446,6 +13460,43 @@ let
...
+    version = "1.15";
+    src = fetchurl {
+      url = mirror://cpan/authors/id/I/IL/ILYAZ/modules/MP3-Tag-1.15.zip;
+      sha256 = "1lanbwv97sfsb7h4vsg1v0dv3yghpz01nf3rzl4a9p3ycgs4ib5a";
+    };
+    buildInputs = [ pkgs.unzip ];
+    propagatedBuildInputs = [ CompressZlib ExtUtilsMakeMaker ];
+    unpackPhase= ''
+      unzip $src
+      cd MP3-Tag-1.15
```
Comment (jtojnar) [inline pkgs/top-level/perl-packages.nix:13487]: Cannot we rely on the default `unpackPhase` and do the substitution in `postPatch`?

File: pkgs/top-level/perl-packages.nix [hunk 19]
```diff
@@ -13446,6 +13460,43 @@ let
...
+      substituteInPlace Makefile.PL --replace "'PL_FILES'" "#'PL_FILES'"
+    '';
+    fixupPhase = ''
+      perl data_pod.PL PERL5LIB:$PERL5LIB
```
Comment (jtojnar) [inline pkgs/top-level/perl-packages.nix:13491]: Maybe also move to `postFixup` so that this does not silently break them (see https://github.com/NixOS/nixpkgs/issues/56785#issuecomment-469034978)?

File: pkgs/top-level/perl-packages.nix [hunk 20]
```diff
@@ -13873,6 +13924,22 @@ let
...
   };
 
+  NetFreeDB = buildPerlPackage {
+    pname = "Net-FreeDB";
+    version = "0.10";
+    src = fetchurl {
+      url = mirror://cpan/authors/id/D/DS/DSHULTZ/Net-FreeDB-0.10.tar.gz;
+      sha256 = "11dfi14qnzsnmr71cygir85zfj15n08b7d5g0i4cj5pb70if2hzp";
+    };
+    #doCheck = false;
```
Comment (jtojnar) [inline pkgs/top-level/perl-packages.nix:13934]: Drop?

File: pkgs/applications/misc/gcstar/default.nix (hunk 10 above)
Comment (dasj19) [reply pkgs/applications/misc/gcstar/default.nix:43]: I left them in `propagatedBuildInputs`

File: pkgs/applications/misc/gcstar/default.nix (hunk 8 above)
Comment (jtojnar) [reply pkgs/applications/misc/gcstar/default.nix:19]: This is still overindented.

File: pkgs/applications/misc/gcstar/default.nix [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+
+  src = fetchFromGitLab {
+      owner = "Kerenoc";
+      repo = "GCstar";
+      rev = "v${version}";
+      sha256 = "1vqfff33sssvlvsva1dflggmwl00j5p64sn1669f9wrbvjkxgpv4";
+    };
+
+  perlDeps = [
+    perl
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:22]: I would worry about coherence and either use `perl` + `perl.pkgs.*`, or `perlPackages.perl` + `perlPackages.*`.

File: pkgs/applications/misc/gcstar/default.nix [hunk 22]
```diff
@@ -0,0 +1,76 @@
+{ stdenv
+, fetchFromGitLab
+, bash
+, perl
+, gtk2
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:5]: bash and gtk2 do not seem to be used.

File: pkgs/applications/misc/gcstar/default.nix (hunk 21 above)
Comment (dasj19) [reply pkgs/applications/misc/gcstar/default.nix:22]: I am afraid I don't get this one... I am not that strong at syntax... do you mean something like

`perlDeps = perl + perlPackages [ package1 package2 .... ]`

(I know that it is syntactically wrong but don't know how to fix it off the top of my head)

File: pkgs/applications/misc/gcstar/default.nix [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+  meta = with stdenv.lib; {
+    homepage = "https://gitlab.com/Kerenoc/GCstar";
+    description = "A general purpose collection manager";
+    longDescription = ''
+      GCstar is an application for managing your collections.
+      It supports many types of collections, including movies, books, games, comics, stamps, coins, and many more.
+      You can even create your own collection type for whatever unique thing it is that you collect!
+      Detailed information on each item can be automatically retrieved from the internet and you can store additional data, such as the location or who you've lent it to. You may also search and filter your collections by many criteria.
+    '';
+    license = licenses.gpl2;
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:72]: Looks like gpl2Plus: https://gitlab.com/Kerenoc/GCstar/-/blob/dde0f1fbcc3c5b6bd68307ed2c3af0d68ea24bdf/gcstar/lib/gcstar/GCSplash.pm#L11-12

File: pkgs/applications/misc/gcstar/default.nix [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+
+    cd gcstar
+
+    perl install --text --prefix=$out
+
+    runHook postInstall
+  '';
+
+  postFixup = ''
+
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:59]:
```suggestion
```

File: pkgs/applications/misc/gcstar/default.nix [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+    wrapProgram $out/bin/gcstar --prefix PERL5LIB : $PERL5LIB
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = "https://gitlab.com/Kerenoc/GCstar";
+    description = "A general purpose collection manager";
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:65]: Maybe we should follow https://gitlab.com/Kerenoc/GCstar/-/blob/Test/gcstar/packages/GCstar_Packaging_Policy.txt

File: pkgs/applications/misc/gcstar/default.nix (hunk 21 above)
Comment (dasj19) [reply pkgs/applications/misc/gcstar/default.nix]: I wrote it this way:
```

perlDeps = [ perl ] ++ (with perlPackages; [
    ArchiveZip
    DateCalc
    DateTimeFormatStrptime
    Glib
    Gtk2
    GD
    GDGraph
    GDText
    HTMLParser
    JSON
    ImageExifTool
    librelative
    LWPUserAgent
    LWPProtocolHttps
    MP3Info
    MP3Tag
    NetFreeDB
    OggVorbisHeaderPurePerl
    Pango
    XMLSimple
    XMLParser
  ]);
```

File: pkgs/applications/misc/gcstar/default.nix (hunk 25 above)
Comment (dasj19) [reply pkgs/applications/misc/gcstar/default.nix]: I took the text from that file. How about displaying it under "Office" in the menus.... how can we control that?

File: pkgs/applications/misc/gcstar/default.nix (hunk 21 above)
Comment (jtojnar) [reply pkgs/applications/misc/gcstar/default.nix]: Sorry, I just meant using `perl` from `perlPackages` or using packages from `perl.pkgs` to avoid mismatches when someone overrides just one of those.

File: pkgs/applications/misc/gcstar/default.nix [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+  version = "1.7.2";
+
+  src = fetchFromGitLab {
+    owner = "Kerenoc";
+    repo = "GCstar";
+    rev = "v${version}";
+    sha256 = "1vqfff33sssvlvsva1dflggmwl00j5p64sn1669f9wrbvjkxgpv4";
+  };
+
+  perlDeps = [ perl ] ++ (with perlPackages; [
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:19]: Grepping nixpkgs, it does not appear that `perlDeps` has any special meaning. Why not just set add these directly to `buildInputs`?

File: pkgs/applications/misc/gcstar/default.nix [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+    MP3Tag
+    NetFreeDB
+    OggVorbisHeaderPurePerl
+    Pango
+    XMLSimple
+    XMLParser
+  ]);
+
+  nativeBuildInputs = [ wrapGAppsHook ];
+  propagatedBuildInputs = perlDeps;
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:44]: Do they need to be propagated?

File: pkgs/applications/misc/gcstar/default.nix (hunk 25 above)
Comment (jtojnar) [reply pkgs/applications/misc/gcstar/default.nix]: That should be the `Category` in the installed `desktop` file.

Comment (dasj19) [reply pkgs/applications/misc/gcstar/default.nix]: The default desktop file has the "Category" office, so no patching is needed.

File: pkgs/applications/misc/gcstar/default.nix (hunk 21 above)
Comment (dasj19) [reply pkgs/applications/misc/gcstar/default.nix:19]: Now it makes use of `perl` from `perlPackages`

File: pkgs/applications/misc/gcstar/default.nix (hunk 26 above)
Comment (dasj19) [reply pkgs/applications/misc/gcstar/default.nix]: Added to `buildInputs`, thanks

File: pkgs/applications/misc/gcstar/default.nix [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+  version = "1.7.2";
+
+  src = fetchFromGitLab {
+    owner = "Kerenoc";
+    repo = "GCstar";
+    rev = "v${version}";
+    sha256 = "1vqfff33sssvlvsva1dflggmwl00j5p64sn1669f9wrbvjkxgpv4";
+  };
+
+  buildInputs = [ wrapGAppsHook ] ++ (with perlPackages; [
```
Comment (jtojnar) [inline pkgs/applications/misc/gcstar/default.nix:18]: `wrapGAppsHook` should go to `nativeBuildInputs`, though.

Comment (dasj19) [reply pkgs/applications/misc/gcstar/default.nix]: Fixed. Thanks

**Comment by dasj19** [comment]
I have been using this for weeks, found no issue with the software... can we release it?
